Assets/UI/Panels/NotificationPanel_CQUI.lua

```diff
@@ -101,18 +121,170 @@ function OnNotificationAdded( playerID:number, notificationID:number )
     BASE_CQUI_OnNotificationAdded(playerID, notificationID);
 end
 
+-- ===========================================================================
+function RegisterHandlers()
+    BASE_CQUI_RegisterHandlers();
+    
+    g_notificationHandlers[NotificationTypes.CIVIC_BOOST].Activate = OnCivicBoostActivateNotification;
+    g_notificationHandlers[NotificationTypes.TECH_BOOST].Activate = OnTechBoostActivateNotification;
+end
+
+
+-- ===========================================================================
+-- CUSTOM NOTIFICATIONS, 2020-09-11 Infixo
+-- Usage:
+-- NotificationManager.SendNotification(iNotifyPlayer, notificationData.Type, msgString, sumString, pPlot:GetX(), pPlot:GetY());
+-- Potential notifications to be added:
+-- city border expands
+-- populations grows
+-- trade deal expired
+-- goody hut reward [there is already NOTIFICATION_DISCOVER_GOODY_HUT]
+
+--[[
+USER_DEFINED_1 - used by BlackDeathScenario and CivRoyaleScenario
+USER_DEFINED_2 - used by BlackDeathScenario and CivRoyaleScenario
+USER_DEFINED_3 - used by BlackDeathScenario and CivRoyaleScenario
+USER_DEFINED_4 - used by BlackDeathScenario and CivRoyaleScenario
+USER_DEFINED_5 - used in CivRoyaleScenario
+USER_DEFINED_6 - used in CivRoyaleScenario
+USER_DEFINED_7 - free => City Border Expands
+USER_DEFINED_8 - free => Population Grows
+USER_DEFINED_9 - free => Trade Deal Expired
+--]]
+
+
+
+-- debug routine - prints a table (no recursion)
+function dshowtable(tTable:table)
+	if tTable == nil then print("dshowtable: table is nil"); return; end
+	for k,v in pairs(tTable) do
+		print(k, type(v), tostring(v));
+	end
+end
+
+-- ===========================================================================
+
+-- prepare reward descriptions in advance, to simplify the process later
+-- default behavior - read Description and fill it with Amount param from the Modifier
+g_RewardExceptions = {
+    -- base
+    GOODYHUT_GRANT_SCOUT      = "desc", -- direct description
+    GOODYHUT_GRANT_UPGRADE    = "desc",
+    GOODYHUT_GRANT_EXPERIENCE = "exp", -- experience
+    GOODYHUT_HEAL             = "desc",
+    GOODYHUT_GRANT_BUILDER    = "unit", -- unit granted
+    GOODYHUT_GRANT_TRADER     = "unit",
+    GOODYHUT_GRANT_SETTLER    = "unit",
+    -- xp2
+    GOODYHUT_GOVERNOR_TITLE   = "desc",
+    GOODYHUT_RESOURCES        = "res", -- resources <Text>[COLOR_FLOAT_MILITARY]+{1_Num} [ICON_{2_Icon}] {3_Resources}[ENDCOLOR]</Text>
+    -- gran colombia maya
+    METEOR_GRANT_GOODIES      = "desc",
+}
+--dshowtable(g_RewardExceptions);
+
+g_RewardDescriptions = {}; -- this is a global variable, so other mods can add its own rewards and hook descriptions in here
+
+function Initialize_RewardDescriptions()
+    --print("Initialize_RewardDescriptions");
+    -- helper to get arguments from modifiers
+    local function GetModifierParam(sModifierID:string, sName:string)
```

Comment:
Scoping. These are local functions, safe to use. It just makes the coding easier e.g. you have access to local parameters, no need to pass everything to the outside. Etc. Plus they should be faster than a global call - but this is hard to verifiy.

Reply:
Btw, same reason for using LL instead of Locale.Lookup. The code has better readability and the local variable is supposed to faster. This was actually thoroughly tested in Civ5 when creating EUI.